### ISSUE TITLE
Filtering results in blank rows

### ISSUE THREAD

**Issue by aaronbeall** [handsontable/handsontable#5208]
### Description
I've had a number of issues with blank rows appearing in my table when there are no blank rows in my source data:

![image](https://user-images.githubusercontent.com/2973900/41798759-e3274cac-763c-11e8-85c9-b6c4b1d13096.png)

### Steps to reproduce
I've had a hard time creating simple steps to reproduce but it happens when filters are involved, sometimes sorting, and I change the source data. The data is an array of objects, not an array of arrays.

### Demo

This isn't quite my setup but it shows blank rows:
http://jsfiddle.net/v4grx6zc/13/

![image](http://g.recordit.co/SqKjJTj9Zb.gif)

In the demo I'm applying the filter via `hot.getPlugin("Filters").filter()` after changing the source data and that causes a blank row to show up (otherwise the row doesn't show up at all), then after a 1s timeout I re-apply the filter and the blank row goes away. In my actual app I don't ever manually apply `filter()` but I do `render()` manually after changing the source data, and that's when the blank rows appear. Oddly in the demo using `hot.render()` after updating the rows doesn't render the new row at all.

### Your environment
* Handsontable version: 4.0.0
* Browser Name and version: Chrome 65
* Operating System: MacOS High Sierra

**Comment by AMBudnik** [comment]
Sorry for a delay @aaronbeall 
I did not see the issue before.

I have updated the demo as it loaded old references http://jsfiddle.net/nf9yv5ph/

**Comment by wszymanski** [comment]
Maybe related to the below, simplified example (rows with indexes `1` and `4` are trimmed):

![screenshot 2019-02-06 at 10 25 21](https://user-images.githubusercontent.com/141330/52332678-af888280-29fb-11e9-8096-1625e02af4ec.png)

**Comment by AMBudnik** [comment]
The issue is replicable in v `7.1.1` http://jsfiddle.net/handsoncode/58k6v1p7/ 
but it works completely different in `8.0.0.-rc1` http://jsfiddle.net/handsoncode/t8nbocg9/

![GIF33](https://user-images.githubusercontent.com/10757813/65045311-daa65d80-d95e-11e9-905e-929d6c353768.gif)

When we open up the filters we do to see any new rows (blank/empty)
![GIF13](https://user-images.githubusercontent.com/10757813/65045443-12150a00-d95f-11e9-871c-885b1984e66d.gif)

In `8.0.0.-rc1` when I change the current button logic to add a new row via a simple `alter` method I get

![image](https://user-images.githubusercontent.com/10757813/65045692-8ea7e880-d95f-11e9-9fcd-04885d3cc2f2.png)

While it worked well in `7.1.1.`
![image](https://user-images.githubusercontent.com/10757813/65045750-a5e6d600-d95f-11e9-8355-ee1a9cb1a953.png)

That's why I'm calling this a regression.

**Comment by wojciechczerniak** [comment]
The examples that are linked modify `data` by reference:
```js
const row = { a: 'First', b: 'bar', c: 'blah', d: 'boo', e: '12/1/2015', f: 100 * Math.random() | 0 };
data[data.length] = row;
```

Without `observeChanges` this never worked well. And now that the `observeChanges` is gone the correct way to do this is to use our API:
```js
  const newRowIndex = hot.countRows();
  hot.alter('insert_row', newRowIndex);
  hot.setDataAtRowProp([
    [newRowIndex, 'a', 'First'],
    [newRowIndex, 'b', 'bar'],
    [newRowIndex, 'c', 'blah'],
    [newRowIndex, 'd', 'boo'],
    [newRowIndex, 'e', '12/1/2015'],
    [newRowIndex, 'f', 100 * Math.random() | 0],
  ]);
```

Issue, with headers is caused by:
```js
hot.updateSettings({
    colHeaders: createColHeaders()
});
```
Not related to index mappers, I think that we should exclude it from this example.

Lets start with this example: https://jsfiddle.net/syd32unz/

Unfortunately, the problem still exists in v8.0.0-rc.

**Comment by wojciechczerniak** [comment]
Ok, we know where the problem is. `setDataAtRowProp` and `populateFromArray` are async so calling `hot.getPlugin("Filters").filter()` does what it can. Rows are created (this is done synchronously) but then data is not available until validation is finished (async). Filter is applied to early.

@aaronbeall I think that it's best to apply the filter after the data was validated and saved. I would move the `hot.getPlugin("Filters").filter()` call to `afterChange` hook with custom `source` to avoid applying filters on every change: https://jsfiddle.net/mjtqLaf7/1/ (and I've changed `colHeaders` to a function, so `updateSettings` and rerender are unnecessary). Sorry, I didn't noticed that earlier, you would solve your issue earlier

We will check if we can do something about `hot.getPlugin("Filters").filter()` results. The rows shouldn't be empty, after the validation they should be filled with the data.

**Comment by AMBudnik** [comment]
I've skipped the code example for the `alter` method in the previous example so I'm attaching a more detailed issue explanation.

The button logic from the example has been changed with

``` js
document.getElementById("add").onclick = e => {
  hot.alter('insert_row',1);
  setTimeout(() => {
    hot.getPlugin("Filters").filter();
  }, 1000)
}
// add new blank row, filter out what doesn't meet the first 
```

All the examples were tested with Chrome 77 and FireFox 69 on Windows 10 (using JSFiddle, not tested locally).

#### `7.1.1` -WORKING
http://jsfiddle.net/handsoncode/us9g5pLy/
![711](https://user-images.githubusercontent.com/10757813/65420007-53defe00-de00-11e9-9ef7-f64448018910.gif)

#### First build for 8.0.0.-rc1`` - BROKEN
http://jsfiddle.net/handsoncode/9eqgx5ma/
![800rc1-first_build](https://user-images.githubusercontent.com/10757813/65420131-addfc380-de00-11e9-813b-a68ad75a7ea1.gif)


#### Newest build (from `23/09/2019`) - **FIXED**. Works like `7.1.1`
http://jsfiddle.net/handsoncode/h651zetn/
![800rc1-new](https://user-images.githubusercontent.com/10757813/65419990-4295f180-de00-11e9-8220-834c0ec6d42d.gif)

**Comment by wszymanski** [comment]
In summary, we have:

- a problem when using `filter` (sync) after `populateFromArray` (async); this shouldn't be used in this way, but still it shouldn't give such effect - so I would categorize this issue still as a bug,
-  flickering headers may occur - bug
- not showing headers after data insertion, mentioned by @AMBudnik - it's fixed, so I take off the `Regression` label.

**Comment by aaronbeall** [comment]
Thanks everyone, this definitely helps.

My real world use-case is very complicated, it's hard to tell the intersection of actions that causes the blank rows to appear in my case. I've gotten it to mostly go away, but there are still some reports from users. I do modify data by reference, I don't use `alter()` or `setDataAtRowProp()` or `populateFromArray()` and I only call `filter()` explicitly _after_ encountering blank rows (as a fix)... I want to switch to using `alter()` instead of modifying data by reference but this gets into my [problems I mentioned here](https://github.com/handsontable/handsontable/issues/5270#issuecomment-541121892).

**Comment by AMBudnik** [comment]
Works well using the official `8.0.0-beta1` http://jsfiddle.net/nuv2og08/

**Comment by aninde** [comment]
Wors well using official `v8` http://jsfiddle.net/aninde/pnj56cv2/.

**Comment by AMBudnik** [comment]
Issue solved in v 8.0.0.


CDN https://cdn.jsdelivr.net/gh/handsontable/handsontable@latest/dist/handsontable.full.css
https://cdn.jsdelivr.net/gh/handsontable/handsontable@latest/dist/handsontable.full.js
NPM npm i handsontable
Release Notes https://github.com/handsontable/handsontable/releases
